components/org.wso2.carbon.identity.api.user.backupcode/org.wso2.carbon.identity.api.user.backupcode.v1/src/main/java/org/wso2/carbon/identity/rest/api/user/backupcode/v1/core/BackupCodeService.java

```diff
@@ -0,0 +1,237 @@
+
+/**
+ *
+ */
+public class BackupCodeService {
+
+    private static final Log log = LogFactory.getLog(BackupCodeService.class);
+
+    public Object initBackupCodes() {
+
+        if (!isValidAuthenticationType()) {
+            throw handleError(Response.Status.FORBIDDEN, USER_ERROR_ACCESS_DENIED_FOR_BASIC_AUTH);
+        }
+
+        BackupCodeResponseDTO backupCodeResponseDTO = new BackupCodeResponseDTO();
+        User user = getUser();
+        try {
+            Map<String, String> claims = BackupCodeGenerator.generateBackupCodes(user.toFullQualifiedUsername(), false);
+            String backupCodes = BackupCodeGenerator.addBackupCodeClaimsAndRetrievingBackupCodes(claims,
+                    user.toFullQualifiedUsername());
+            backupCodeResponseDTO.setBackupCodes(new ArrayList<>(Arrays.asList(backupCodes.split(","))));
+            return backupCodeResponseDTO;
+        } catch (BackupCodeException e) {
+            throw handleException(e, SERVER_ERROR_GENERAL);
+        }
+    }
+
+    public Object refreshBackupCodes() {
+
+        if (!isValidAuthenticationType()) {
+            throw handleError(Response.Status.FORBIDDEN, USER_ERROR_ACCESS_DENIED_FOR_BASIC_AUTH);
+        }
+
+        BackupCodeResponseDTO backupCodeResponseDTO = new BackupCodeResponseDTO();
+        User user = getUser();
+        try {
+            Map<String, String> claims = BackupCodeGenerator.
+                    generateBackupCodes(user.toFullQualifiedUsername(), true);
+            String backupCodes = BackupCodeGenerator.addBackupCodeClaimsAndRetrievingBackupCodes(claims,
+                    user.toFullQualifiedUsername());
+            backupCodeResponseDTO.setBackupCodes(new ArrayList<>(Arrays.asList(backupCodes.split(","))));
+        } catch (BackupCodeException e) {
+            throw handleException(e, SERVER_ERROR_GENERAL);
+        }
+        return backupCodeResponseDTO;
+    }
+
+    public APIError handleInvalidInput(BackupCodeConstants.ErrorMessage errorEnum, String... data) {
+
+        return handleError(Response.Status.HTTP_VERSION_NOT_SUPPORTED, errorEnum);
+    }
+
+    public void deleteBackupCodes() {
+
+        if (!isValidAuthenticationType()) {
+            throw handleError(Response.Status.FORBIDDEN, USER_ERROR_ACCESS_DENIED_FOR_BASIC_AUTH);
+        }
+
+        User user = getUser();
+        try {
+            BackupCodeGenerator.deleteBackupCodes(user.toFullQualifiedUsername());
+        } catch (AuthenticationFailedException e) {
+            throw handleException(e, USER_ERROR_UNAUTHORIZED_USER);
+        } catch (BackupCodeException ex) {
+            throw handleException(ex, SERVER_ERROR_RETRIEVING_REALM_FOR_USER);
+        }
+    }
+
+    /**
+     * Get authenticated user.
+     *
+     * @return Authenticated user
+     */
+    public static User getUser() {
+
+        return ContextLoader.getUserFromContext();
+    }
+
+    private boolean isValidAuthenticationType() {
```

Comment:
Verify the logic of this method

Reply:
The logic here is valid. The below comment explain the requirement of this validation
```
        /*
         * Check whether the request is authenticated with basic auth. Backup code endpoint should not be allowed for
         * basic authentication. This approach can be improved by providing a Level of Assurance (LOA) and
         * checking that in BackupCodeService.
         */
```